### PR TITLE
Variance extractor

### PR DESCRIPTION
**Pull request by ctoennis**
This pull request is to add a new ImageExtractor subclass to generate images from the variance of waveforms that will be needed for the pointing calibration using stars. This request is related to issue #2542 .

### CODE REVIEW

File: src/ctapipe/image/extractor.py
Comment (TjarkMiener) [inline src/ctapipe/image/extractor.py:1310]: I think the DL1CameraContainer should be adjusted rather than storing the variances where the charge images belong and setting peak_time to zero. This will introduce breaking changes ...

Comment (maxnoe) [reply src/ctapipe/image/extractor.py:1310]: peak_time should be `None`, not 0, which will e.g. make it so that this field is skipped when writing out.

If you need to give a value, it needs to have the correct shape

File: src/ctapipe/image/extractor.py
Comment (maxnoe) [inline src/ctapipe/image/extractor.py:1309]: Do we really just want to compute the variance of the hole waveform?

I know pedvar as the the variance over an integration window. 

You place as many integration windows as you can in the waveform and then compute the variance over the integrals.

File: src/ctapipe/image/extractor.py
Comment (mexanick) [reply src/ctapipe/image/extractor.py:1310]: Do we want to add a metadata field that describes the meaning of the image? Could be an enum of "charge", "variance" etc. with the default value being "charge" for backwards compatibility

File: src/ctapipe/image/extractor.py
Comment (mexanick) [reply src/ctapipe/image/extractor.py:1309]: that might be interesting, however, I didn't test it... A simple variance like here works for sure. 
@ctoennis I think the idea of Max is nice and we can implement it alongside and compare later

Comment (maxnoe) [reply src/ctapipe/image/extractor.py:1309]: Just to be clear, that's not my idea but comes from the original ped var paper, I think by VERITAS, let me dig it up

**Comment by kosack** [comment]
If I understand correctly, what you are basically doing here is computing an event-wise pedestal variance. In the data model (see the original top-level data model document), there are a few places you can put pedestals:

- _dl1.event.telescope.pedestal_: pedestals measured for each event (usually in a time window as @maxnoe mentioned, since you can for example just use time slices outside the trace integration window, even for Cherenkov events). This is also computed by some cameras directly in hardware as dl0 information.
- _dl1.monitoring.telescope.pedestal_: is the pedestal computed from an ensemble of pedestal events, or from Cherenkov events with the shower cut out  (this is also in the DL0.monitoring.telescope.pedestal, computed by the cameras) computed every few seconds.
- dl1.service.telescope.pedestal: pedestals measured from some standard run that are applied for e.g. RTA

So I think what you are computing is the first one, and you should store it as pedestal information:  It is in fact a second measurement of the pedestal and its variance, and could be compared to the other methods (even if the use case you are describing here is for doing star tracking, star tracking is computed from the pedestal variances, which can be over many time scales)

**Comment by fwerner** [comment]
In addition to what has been said I'd suggest to clarify the purpose of this extractor: it is calculating the waveform _sample_ variance, which should be reflected in the name/docstring to communicate that it has a complicated relationship to charge.

File: src/ctapipe/image/extractor.py
Comment (maxnoe) [inline src/ctapipe/image/extractor.py:1302]: First line needs to be a single line summary followed by an empty line, then you can give more details.

I would propose:

```
Compute variance over samples for each waveform.
```

File: src/ctapipe/image/extractor.py
Comment (ctoennis) [reply src/ctapipe/image/extractor.py:1310]: I will see how well this works compared to just looking at the variance directly.

**Comment by ctoennis** [comment]
The thing we want to calculate is the variance of the waveforms specifically of interleaved pedestal events and not the variance of the pedestal in units of charge. 

@kosack : What we get is a sort of image based on the variance of the waveform in each pixel. What we get is not a pedestal value, so putting it in a pedestal container is not the best option i think. 

I think it is best to write a container appropriate for these variance images similar to the DL1CameraContainer

File: src/ctapipe/image/extractor.py
Comment (mexanick) [reply src/ctapipe/image/extractor.py:1310]: since the amplitude of the signal in bin N+1 does not depend on the amplitude of the signal in bin N, then the variance of a sample and the variance of charge are equivalent (cov(Ai, Aj) == 0). The only differences are statistical error on the calculated variance and the normalization.

Comment (fwerner) [reply src/ctapipe/image/extractor.py:1310]: I'm not sure I understand the context, so apologies if I misunderstood what you've written - just wanted to say that if by "bin" you mean "time slice" for a given channel: consecutive samples are _not_ independent because the single-p.e. pulse shapes are several nanoseconds long (tens of nanoseconds for some cameras). Also electronics noise is usually not white.

Comment (mexanick) [reply src/ctapipe/image/extractor.py:1310]: You understand it correctly and I was thinking of what you wrote. However we consider NSB radiation, which is commonly modeled as a Poisson process, i.e. the numbers of photons hitting the pixel in two consecutive time intervals are independent. Now, for the PMT-equipped camera the conversion of photons to p.e. will also be uncorrelated. This is not correct for SiPM cameras, where generalized Poisson shall be applied.

**Comment by ctoennis** [comment]
I just added a custom continer for the variance images. I also compared the variance calculation using just a plain variance and a variance of integrated chunks of the waveforms using the FixedWindowSum extractor in the plot below.
comparemethod

The x axis shows the number of images used in calculating an average variance value that is shown on the y axis. 

Though the methods are generally statistically equivalent the average variance becomes stable with fewer images when using a simple variance over the variance from integrated chunks.

**Comment by maxnoe** [comment]
> Though the methods are generally statistically equivalent the average variance becomes stable with fewer images when using a simple variance over the variance from integrated chunks.

That's not surprising, since you have just more samples. But you change *of what you compute the variance*

**Comment by ctoennis** [comment]
Hi all, can we restart the discussion here? I made a small update to the extractor to contain timing information that will later be needed in the corresponding subclass of StatisticsExtractor that i am working on with Tjark. 

There was a decision needed as to where the variance images will be attached to. Should this find a place somewhere in dl1.event.telescope.pedestal or something similar?

File: src/ctapipe/image/extractor.py
Comment (mexanick) [inline src/ctapipe/image/extractor.py]: This has to be a _single_ line, really concise. A longer description can follow after one blank line.

Comment (ctoennis) [reply src/ctapipe/image/extractor.py]: sorry about that. i shortened it.

**Comment by maxnoe** [comment]
Hi @ctoennis the lint step in the CI is failing due to some typos and wrong formatting.

To check locallly, make sure to install the pre-commit hooks:
https://ctapipe.readthedocs.io/en/latest/developer-guide/getting-started.html#installing-ctapipe-in-development-mode

**Comment by ctoennis** [comment]
i fixed all the isses with the spelling, code checks etc.

File: src/ctapipe/containers.py
Comment (maxnoe) [inline src/ctapipe/containers.py:176]: SIMPLE and SAMPLE are extremely similar and `SAMPLE` could also mean "variance over the waveform samples" so what you refer to here as "SIMPLE".

Maybe better `SAMPLE` vs. `INTEGRATED` ?

Also please add a color before the comment: `#: ` which is the special syntax for docstrings for member attributes.

File: src/ctapipe/containers.py
Comment (maxnoe) [inline src/ctapipe/containers.py:535]: I'd call this `variance`, not `image` if it is allways the variance

File: src/ctapipe/containers.py
Comment (maxnoe) [inline src/ctapipe/containers.py:541]: Why do we need to store the time here in this container?

File: src/ctapipe/image/extractor.py
Comment (maxnoe) [inline src/ctapipe/image/extractor.py:1310]: Is this really an image extractor? It doesn't follow the abstract API of `ImageExtractor`, it returns a different data type and it's arguments are also different.

For it to *be* an image extractor, it must be usable if configured through the configuration system.

This fails for example:

```
❯ ctapipe-process -i dataset://gamma_prod5.simtel.zst -o /tmp/test.dl1.h5 --CameraCalibrator.image_extractor_type=VarianceExtractor

2024-07-10 15:28:47,988 WARNING [ctapipe.core.telescope_component] (telescope_component.attach_subarray): TelescopeParameter type argument 'SST_1M_*' did not match any known telescope types
2024-07-10 15:28:49,210 ERROR [ctapipe.ctapipe-process] (tool.run): Caught unexpected exception: VarianceExtractor.__call__() got an unexpected keyword argument 'selected_gain_channel'
Traceback (most recent call last):
  File "/home/mnoethe/Uni/CTA/ctapipe/src/ctapipe/core/tool.py", line 426, in run
    self.start()
  File "/home/mnoethe/Uni/CTA/ctapipe/src/ctapipe/tools/process.py", line 311, in start
    self.calibrate(event)
  File "/home/mnoethe/Uni/CTA/ctapipe/src/ctapipe/calib/camera/calibrator.py", line 329, in __call__
    self._calibrate_dl1(event, tel_id)
  File "/home/mnoethe/Uni/CTA/ctapipe/src/ctapipe/calib/camera/calibrator.py", line 277, in _calibrate_dl1
    dl1 = extractor(
          ^^^^^^^^^^
TypeError: VarianceExtractor.__call__() got an unexpected keyword argument 'selected_gain_channel'
```

**Comment by ctoennis** [comment]
i can remove trigger time. But i am not sure how to proceed on the difference to the ImageExtractor API. I could make it that the variance image is stored in a DL1ImageContainer and that it complies with the API or make it a separate class that doesn't inherit from the ImageExtractor. 

I think what i generate here is a type of image as i try to look for stars in the camera FOV with them. But if it makes more sense to you it can be separate from the ImageExtractors.

**Comment by maxnoe** [comment]
The question is if it is possible to adhere to the API that is expected from the image extractors: if yes, no problem, let it be an image extractor. But that means for example the call above should work and correctly write out the data to the HDF5 file.

If you need the different API, it should be its own thing and we need to think about how to integrate it into the process tool for example.

**Comment by kosack** [comment]
> The question is if it is possible to adhere to the API that is expected from the image extractors: if yes, no problem, let it be an image extractor. But that means for example the call above should work and correctly write out the data to the HDF5 file.

I think what this class does is not the same as an `ImageExtractor`, or am I missing something? `ImageExtractors` take a "movie" (a set of waveforms per pixel for one event) for a single event and convert them to  set of time-independent dependent images that must include the  _peak time_ and _integral_ over a window, but can also optionally inlcude any other parameters of the waveform like the pulse width or variabce.    Generally, there is only one ImageExtractor that is run at a time.    That doesn't seem to be what you have here.

**Comment by maxnoe** [comment]
> I think what this class does is not the same as an ImageExtractor, or am I missing something?

You could think of this as an `ImageExtractor` that uses the variance of each waveform to produce the image and doesn't produce a `peak_time`.

**Comment by mexanick** [comment]
I believe, it is an ImageExtractor, the `__call__()` signature is of course wrong here. It should copy the one of the base extractor:
```
@abstractmethod
    def __call__(
        self, waveforms, tel_id, selected_gain_channel, broken_pixels
    ) -> DL1CameraContainer:
```

Only `waveforms` are then used, the other parameters are simply ignored.

The only issue, that has to be resolved, is the return type. `DL1CameraContainer` in its current shape contains a number of fields that are not needed in this case (actually, everything except `image`). 

I'd propose to add an intermediate container:

```python
class CameraImageContainer(Container):
    """
    Storage of output of generic camera images (any reduction of waveforms).
    """

    image = Field(
        None,
        "Numpy array of camera image, after waveform extraction."
        "Shape: (n_pixel) if n_channels is 1 or data is gain selected"
        "else: (n_channels, n_pixel)",
    )

class DL1CameraContainer(CameraImageContainer):
    """
    Storage of output of camera calibration e.g the final calibrated
    image in intensity units and the pulse time.
    """
    peak_time = Field(
        None,
        "Numpy array containing position of the peak of the pulse as determined by "
        "the extractor."
        "Shape: (n_pixel) if n_channels is 1 or data is gain selected"
        "else: (n_channels, n_pixel)",
    )
    # rest of the DL1CameraContainer fields
    
class VarianceContainer(CameraImageContainer):
    """
    Storage of output of camera variance image.

    Stores the variance of waveform in each pixel, composed as an image.
    """
    variance_method = Field(
        VarianceType.SIMPLE,
        "Method by which the variance was calculated"
        "This can either be a plain variance"
        "or a variance of integrated samples",
        type=VarianceType,
    )
    # I don't think you need anything else here. The trigger time you extract from elsewhere in the event tree.
```
and change the return type of base `ImageExtractor` to `CameraImageContainer`

**Comment by maxnoe** [comment]
If it's just that the `VarianceExtractor` doesn't provide a `peak_time`, I think that's fine, just leave it `None` (the default value).

Adding an additional field to the `DL1CameraContainer` that's an "image_type" sounds like a good idea.

I would avoid having extractors that have different return types.

**Comment by mexanick** [comment]
Following a private discussion with @maxnoe , the new Container class is not needed. 
@ctoennis just use the `DL1CameraContainer` as a return type, setting the extractor type as a container meta keyword. You only need to fill `Image` field and leave all others as defaults (`None`). In this case, the serialization will be done correctly and only the image that you need will be stored. I.e. something like this (assuming you change VarianceType enum to WAVEFORM):

```python
def __call__(
        self, waveforms, tel_id, selected_gain_channel, broken_pixels
    ) -> DL1CameraContainer:
    container = DL1CameraContainer(image=np.nanvar(waveforms, dtype="float32", axis=2))
    container.meta["ExtractionMethod"] = str(VarianceType.WAVEFORM)
    return container
```

**Comment by maxnoe** [comment]
Since this is also somewhat connected to #1243 

I.e. we could introduce a new class e.g. `WaveformProperties` that runs after the `CameraCalibrator` and computes stuff like variance, FWHM, etc.

File: src/ctapipe/containers.py
Comment (mexanick) [inline src/ctapipe/containers.py:480]: Why this is removed?

File: docs/changes/2543.feature.rst
Comment (mexanick) [inline docs/changes/2543.feature.rst:2]: This needs an update

File: src/ctapipe/containers.py
Comment (ctoennis) [reply src/ctapipe/containers.py:480]: sorry, that was by accident

File: src/ctapipe/image/tests/test_extractor.py
Comment (maxnoe) [inline src/ctapipe/image/tests/test_extractor.py:430]:
```suggestion
    if Extractor is in (FullWaveformSum, VarianceExtractor):
```

File: src/ctapipe/image/tests/test_extractor.py
Comment (maxnoe) [inline src/ctapipe/image/tests/test_extractor.py:720]: I'd remove this here and just skip the `dl1.image.dtype` check in case it's the `VarianceExtractor`

**Review by maxnoe — DISMISSED** [review]

**Comment by mexanick** [comment]
> Since this is also somewhat connected to #1243
> 
> I.e. we could introduce a new class e.g. `WaveformProperties` that runs after the `CameraCalibrator` and computes stuff like variance, FWHM, etc.

While this might be a good idea, I'd save it for a separate refactoring PR and merge the current (small) addition as is.

**Review by mexanick — DISMISSED** [review]

File: src/ctapipe/image/extractor.py
Comment (maxnoe) [inline src/ctapipe/image/extractor.py:1309]:
```suggestion
    """Calculate the variance over samples in each waveform."""
```

**Review by maxnoe — APPROVED** [review]

**Review by maxnoe — DISMISSED** [review]

**Comment by ctoennis** [comment]
So, is this request then good to merge?

**Comment by maxnoe** [comment]
@ctoennis From my side yes, but we require two approvals and @kosack is on vacation right now. Since he had comments / requests for changes before, I'd like to wait for his approval here.

**Comment by kosack** [comment]
I'll take a look tomorrow or today hopefully! Might take a bit of time since there are a lot of changes.

**Review by kosack — CHANGES_REQUESTED** [review]
Looks good. but I would like to have a test that checks for correctness.  It is a simple algorithm, but the test will also ensure that nobody breaks it in the future. E.g. a test that runs the extractor on some dummy data with a known variance.

**Comment by ctoennis** [comment]
i added a test to check that an actual variance is being calculated.

File: src/ctapipe/image/tests/test_extractor.py
Comment (maxnoe) [inline src/ctapipe/image/tests/test_extractor.py:289]: The shape here is a bit weird for testing an image extractor. Of course for the test this doesn't really matter, but what you are doing here means

n_channels = 5000
n_pixels = 2 
n_samples = 1855

File: src/ctapipe/image/tests/test_extractor.py
Comment (maxnoe) [inline src/ctapipe/image/tests/test_extractor.py:293]: Use `np.testing.assert_all_close(variance, np.var(...))` for better error messages in case of failure

**Review by maxnoe — APPROVED** [review]

**Pull request merged**
